text/3536-unsized-thin-pointers.md

```diff
@@ -0,0 +1,286 @@
+
+```rust
+#[repr(C, align(8))]
+struct SizedRequest {
+	size: u32,
+	id: u16,
+	flags: u16,
+	data: [u8; 1024],
+}
+
+// Compiler error: `impl DynSized` on a type that isn't `!Sized`.
+unsafe impl core::mem::DynSized for SizedRequest {
+	fn size_of_val(&self) -> usize {
+		usize::try_from(self.size).unwrap_or(usize::MAX)
+	}
+}
+```
+
+# Reference-level explanation
+[reference-level-explanation]: #reference-level-explanation
+
+The `core::mem::DynSized` trait acts as a signal to the compiler that the
+size of a value can be computed dynamically by the user-provided trait
+implementation. If references to that type would otherwise be of the layout
+`(ptr, usize)` due to being `!Sized`, then they can be reduced to `ptr`.
+
+The `DynSized` trait does not _guarantee_ that a type will have thin pointers,
+it merely enables it. This definition is intended to be compatible with RFC
+2580, in that types with complex pointer metadata would continue to have fat
+pointers. Such types may choose to implement `DynSized` by extracting their
+custom pointer metadata from `&self`.
+
+Implementing `DynSized` does not affect alignment, so the questions of how to
+handle unknown alignments of RFC 1861 `extern type` DSTs do not apply.
+
+In current Rust, a DST used as a `struct` field must be the final field of the
+`struct`. This restriction remains unchanged, as the offsets of any fields after
+a DST would be impossible to compute statically.
+- This also implies that any given `struct` may have at most one field that
+  implements `DynSized`.
+
+A `struct` with a field that implements `DynSized` will also implicitly
+implement `DynSized`. The implicit implementation of `DynSized` computes the
+size of the struct up until the `DynSized` field, and then adds the result of
+calling `DynSized::size_of_val()` on the final field.
+- This implies it's not permitted to manually `impl DynSize` for a type that
+  contains a field that implements `DynSize`.
+
+# Drawbacks
+[drawbacks]: #drawbacks
+
+## Mutability of value sizes
+
+If the size of a value is stored in the value itself, then that implies it can
+change at runtime.
+
+```rust
+struct MutableSize { size: usize }
+unsafe impl core::mem::DynSized for MutableSize {
+	fn size_of_val(&self) -> usize { self.size }
+}
+
+let mut v = MutableSize { size: 8 };
+println!("{:?}", core::mem::size_of_val(&v)); // prints "8"
+v.size = 16;
+println!("{:?}", core::mem::size_of_val(&v)); // prints "16"
+```
+
+There may be existing code that assumes `size_of_val()` is constant for a given
+value, which is true in today's Rust due to the nature of fat pointers, but
+would no longer be true if `size_of_val()` is truly dynamic.
+
+Alternatively, the API contract for `DynSized` implementations could require
+that the result of `size_of_val()` not change for the lifetime of the allocated
+object. This would likely be true for nearly all interesting use cases, and
+would let `DynSized` values be stored in a `Box`.
```

Comment:
Nothing written here would prevent DynSized values being stored in Box (either via coercion or from_raw) so this does need to work or be banned somehow.

Reply:
After giving it some thought, `Box<T: DynSized>` should probably be disallowed. It would be *convenient* to be able to box a dynamically-sized value as-is, but requiring the size to be immutable for the lifetime of the object would prevent useful operations (for example editing an IPv4 packet in place to drop IP routing options).

Users that need to alloc+free a `DynSized` would need to use a buffer type that manages the actual total capacity and has a fixed `Layout`, and that type could go into a `Box`.